docs(x/ecocredit): add basket and update content (#923)
regen-network/regen-ledger: x/ecocredit/spec/01_concepts.md

# Concepts

## Ecocredit Core

### Credit Class

> A credit class is the primary abstraction for an ecosystem service credit. A credit class is associated with a credit type that represents the primary unit of measurement for the accepted methodologies included within the credit class. Information about the approved methodologies are typically stored off-chain, securely hashed, timestamped, and then included in the metadata field.

Author: Slightly modified and accepted. For future reference, "off-chain data" (hyphen) and "stored off chain" (no hyphen) as well as "hash-based" (hyphen). I kept the link but we should update to point to concepts in the data module once we have them.

For more information about the properties of a project, see [ProjectInfo](https://buf.build/regen/regen-ledger/docs/main/regen.ecocredit.v1#regen.ecocredit.v1.ProjectInfo).

### Credit Type

A credit type is the primary unit of measurement used by the approved methodologies defined within a credit class. A credit type includes a name (e.g. carbon, biodiversity), an abbreviation (a set of 1-3 uppercase characters), a measurement unit (e.g. kilograms, tons), and a decimal precision.

The credit type abbreviation is used to construct the credit class ID. For example, `C01` is the ID for the first credit class that uses the `carbon` credit type where `C` is the credit type abbreviation and `01` is the sequence number (i.e. the first credit class for the given credit type).

Credit types are listed as an on-chain parameter. Adding a new credit type to the list of allowed credit types requires a parameter change proposal.

For more information about the properties of a credit type, see [CreditType](https://buf.build/regen/regen-ledger/docs/main/regen.ecocredit.v1#regen.ecocredit.v1.CreditType).

### Credit Class Creator Allowlist

> The ecocredit module supports the option of restricting credit class creation to a set of addresses. The option to enable the credit class creator allowlist and the list of allowed credit class creators are both on-chain parameters that can only be updated through parameter change proposals.

Reviewer: ```suggestion
The ecocredit module supports the option of restricting credit class creation to a set of addresses. Enabling this restriction and maintaining the list of allowed credit class creators are both controlled via on-chain governance and can only be updated through parameter change proposals.
```

### Credit Class Admin

> The credit class admin is defined within a credit class as the address with the authority to update the credit class (i.e. to update the admin, the list of approved issuers, and the metadata). When a credit class is created, the admin is initially set to the address that created the credit class.

Reviewer: ```suggestion
The credit class admin is the address with the authority to update a given credit class (i.e. to update the admin, the list of approved issuers, and the metadata). When a credit class is created, the admin is initially set to the address that created the credit class.
```

### Credit Class Issuers

> The credit class issuers are defined within a credit class as the addresses with the authority to issue credit batches under the credit class. The list of credit class issuers is defined at the time the credit class is created and only the admin can update the list after the credit class is created.

Reviewer: ```suggestion
The issuers of a given credit class are the addresses with the authority to issue credit batches under that credit class. The list of credit class issuers is set at the time the credit class is created and only the admin can update the list after the credit class is created.
```

### Credit Batch

> Credits are issued in batches by credit issuers granted the authority to issue credits for a given credit class. A credit batch refers to a batch of credits issued at a single point in time.

Reviewer: ```suggestion
All credits are issued in discrete batches by credit issuers. A credit batch refers to a batch of credits issued at a single point in time for a given credit class and project.
```

Tradable credits are credits that the owner has full control over. Tradable credits can be transferred by the owner to another account, put into a basket in return for basket tokens (see [basket](#basket-submodule) for more information), or listed for sale in the [marketplace](#marketplace-submodule), placing the credits in escrow until the sell order is either processed or cancelled (or the credit amount within the sell order is updated). Tradable credits can be retired or cancelled by the owner at any time.

### Retired Credits

Retiring a credit is equivalent to burning a token with the exception that retired credits are actively tracked after they are retired. Retiring a credit implies the owner of the credit is consuming it as an offset. Credits can be retired upon issuance, upon transfer, upon being taken from a [basket](#basket-submodule), upon being sold in the [marketplace](#marketplace-submodule), or directly by the owner. Retiring a credit is permanent.

### Cancelled Credits

Cancelled credits are credit that cannot be transferred or retired. Credits are cancelled in the event that the credit has moved to another registry.

## Basket Submodule

### Basket

> A basket is an abstraction for different types of credits that meet a defined criteria. Credits can be put into a basket in exchange for an equivalent amount of basket tokens. The basket criteria can be set to only accept specific credit types, credit classes, and credit batches that meet a specific date criteria (e.g. credit batches with a minimum start date, credit batches with a start date within a duration of time, or credit batches with a start date year within a number of years into the past). Basket tokens can be returned to the basket at any time in exchange for the equivalent amount of credits.

Reviewer 2: in the current impl, only one credit type can be specified 
```suggestion
A basket is an abstraction for different types of credits that meet a defined criteria. Credits can be put into a basket in exchange for an equivalent amount of basket tokens. The basket criteria can be set to only accept a specific credit type, credit classes, and credit batches that meet a specific date criteria (e.g. credit batches with a minimum start date, credit batches with a start date within a duration of time, or credit batches with a start date year within a number of years into the past). Basket tokens can be returned to the basket at any time in exchange for the equivalent amount of credits.
```

### Direct Buy Order

A direct buy order is an order to buy credits from a specific sell order. The buyer specifies the quantity of credits to buy. If partial fills are enabled in the sell order, the buyer can specify a quantity less than the available quantity. If partial fills are disabled, the buyer must specify the total quantity of credits. The buyer has the option to disable auto-retirement only if the sell order has also disabled auto-retirement. If auto-retirement is enabled, the credits are retired upon being sent to the buyer.

### Allowed Denom

An allowed denom is a token denomination that has been approved to use within a sell order. A seller can only provide an ask price in a token denomination that is included in an on-chain parameter set through an on-chain governance process.

For more information about the properties of an allowed denom, see [AllowedDenom](https://buf.build/regen/regen-ledger/docs/main/regen.ecocredit.marketplace.v1#regen.ecocredit.marketplace.v1.AllowedDenom).

<br/>

---

<br/>

> ![Ecocredit Types](./assets/types.png)

Reviewer: can this be moved to the top of this page?

Author: I actually think we should keep these at the bottom of the page for now as they don't align 1-to-1 with the concepts and they are becoming outdated. I was considering removing them but I think they still provide some value.

> ![Credit Class Roles](./assets/roles.png)

Reviewer: can this be moved right under credit class admin & credit class issuers section? Maybe change those to subheadings, and call the parent section "Credit Class Roles" ?

Reviewer: Also maybe move allowlist (below) image to be right under the description of hte allow list.

Reviewer: Also maybe move allowlist (below) image to be right under the description of hte allow list.
